File: libs/sdk-java/src/main/java/io/nightona/sdk/VolumeService.java
// Copyright Nightona Platforms Inc.
// SPDX-License-Identifier: Apache-2.0

package io.nightona.sdk;

import io.nightona.api.client.api.VolumesApi;
import io.nightona.api.client.model.CreateVolume;
import io.nightona.sdk.model.Volume;

import java.util.List;
import java.util.ArrayList;

/**
 * Service for managing Nightona Volumes.
 *
 * <p>Volumes provide persistent shared storage that can be mounted into Sandboxes.
 */
public class VolumeService {
    private final VolumesApi volumesApi;

    VolumeService(VolumesApi volumesApi) {
        this.volumesApi = volumesApi;
    }

    /**
     * Creates a new volume.
     *
     * @param name volume name
     * @return created {@link Volume}
     * @throws io.nightona.sdk.exception.NightonaException if creation fails
     */
    public Volume create(String name) {
        io.nightona.api.client.model.VolumeDto volumeDto = ExceptionMapper.callMain(
                () -> volumesApi.createVolume(new CreateVolume().name(name), null)
        );
        return toVolume(volumeDto);
    }

    /**
     * Lists all accessible volumes.
     *
     * @return list of available volumes
     * @throws io.nightona.sdk.exception.NightonaException if the API request fails
     */
    public List<Volume> list() {
        List<io.nightona.api.client.model.VolumeDto> volumes = ExceptionMapper.callMain(() -> volumesApi.listVolumes(null, null));
        List<Volume> result = new ArrayList<Volume>();
        if (volumes != null) {
            for (io.nightona.api.client.model.VolumeDto volume : volumes) {
                result.add(toVolume(volume));
            }
        }
        return result;
    }

    /**
     * Retrieves a volume by name.
     *
     * @param name volume name
     * @return matching {@link Volume}
     * @throws io.nightona.sdk.exception.NightonaException if no volume is found or request fails
     */
    public Volume getByName(String name) {
        io.nightona.api.client.model.VolumeDto volumeDto = ExceptionMapper.callMain(() -> volumesApi.getVolumeByName(name, null));
        return toVolume(volumeDto);
    }

    /**
     * Deletes a volume by ID.
     *
     * @param id volume identifier
     * @throws io.nightona.sdk.exception.NightonaException if deletion fails
     */
    public void delete(String id) {
        ExceptionMapper.runMain(() -> volumesApi.deleteVolume(id, null));
    }

    private Volume toVolume(io.nightona.api.client.model.VolumeDto source) {
        Volume volume = new Volume();
        if (source != null) {
            volume.setId(source.getId());
            volume.setName(source.getName());
            volume.setState(source.getState() == null ? null : source.getState().getValue());
        }
        return volume;
    }
}

File: libs/sdk-ruby/lib/nightona/volume_service.rb
# Copyright Daytona Platforms Inc.
# SPDX-License-Identifier: Apache-2.0

# frozen_string_literal: true

module Nightona
  class VolumeService
    include Instrumentation

    # Service for managing Nightona Volumes. Can be used to list, get, create and delete Volumes.
    #
    # @param volumes_api [NightonaApiClient::VolumesApi]
    # @param otel_state [Nightona::OtelState, nil]
    def initialize(volumes_api, otel_state: nil)
      @volumes_api = volumes_api
      @otel_state = otel_state
    end

    # Create new Volume.
    #
    # @param name [String]
    # @return [Nightona::Volume]
    def create(name) = Volume.new(volumes_api.create_volume(NightonaApiClient::CreateVolume.new(name:)))

    # Delete a Volume.
    #
    # @param volume [Nightona::Volume]
    # @return [void]
    def delete(volume) = volumes_api.delete_volume(volume.id)

    # Get a Volume by name.
    #
    # @param name [String]
    # @param create [Boolean]
    # @return [Nightona::Volume]
    def get(name, create: false)
      Volume.new(volumes_api.get_volume_by_name(name))
    rescue NightonaApiClient::ApiError => e
      raise unless create && e.code == 404 && e.message.include?("Volume with name #{name} not found")

      create(name)
    end

    # List all Volumes.
    #
    # @return [Array<Nightona::Volume>]
    def list
      volumes_api.list_volumes.map { |volume| Volume.new(volume) }
    end

    instrument :create, :delete, :get, :list, component: 'VolumeService'

    private

    # @return [NightonaApiClient::VolumesApi]
    attr_reader :volumes_api

    # @return [Nightona::OtelState, nil]
    attr_reader :otel_state
  end
end
